packages/in-flight/src/in-flight.js

```diff
@@ -0,0 +1,67 @@
+const cuid = require('@bugsnag/cuid')
+
+const FLUSH_POLL_INTERVAL_MS = 50
+const inFlightRequests = new Map()
+
+const noop = () => {}
+
+module.exports = {
+  trackInFlight (client) {
+    const originalNotify = client._notify
+
+    client._notify = function (event, onError, callback = noop) {
+      const id = cuid()
+      inFlightRequests.set(id, true)
+
+      const _callback = function () {
+        inFlightRequests.delete(id)
+        callback.apply(null, arguments)
+      }
+
+      client._depth += 1
+
+      try {
+        originalNotify.call(client, event, onError, _callback)
+      } finally {
+        client._depth -= 1
+      }
+    }
+
+    const delivery = client._delivery
+    const originalSendSession = delivery.sendSession
+
+    delivery.sendSession = function (session, callback = noop) {
+      const id = cuid()
+      inFlightRequests.set(id, true)
+
+      const _callback = function () {
+        inFlightRequests.delete(id)
+        callback.apply(null, arguments)
+      }
+
+      originalSendSession.call(delivery, session, _callback)
+    }
+  },
+
+  flush (timeoutMs) {
+    return new Promise(function (resolve, reject) {
+      const timeout = setTimeout(
+        () => { reject(new Error(`flush timed out after ${timeoutMs}ms`)) },
+        timeoutMs
+      )
+
+      const resolveIfNoRequests = function () {
+        if (inFlightRequests.size === 0) {
+          clearTimeout(timeout)
+          resolve()
+
+          return
+        }
+
+        setTimeout(resolveIfNoRequests, FLUSH_POLL_INTERVAL_MS)
```

Comment:
I'm not sure how important this is, because I'm pretty sure `Promise` internally guarantees that it will only call either a resolve or reject callback once…

However, for clean/completeness' sake I think you might want to hold a reference to the poll timeout, and clear it in the in the callback for the timeout on line 49.

The hypothetical situation goes like so:

- at `t=timeoutMs` there are still requests in flight
- the timeout callback fires and `reject()` is called
- at `t>timeoutMs && t<timeoutMs + FLUSH_POLL_INTERVAL_MS` before the process manages to shutdown, the in-flight requests finish
- the `resolveIfNoRequests` callback calls `resolve()` even though `reject()` was already called

Reply:
The ES spec does guarantee that promises can only be resolved/rejected once, but I think it makes sense to do just in case a JS engine isn't totally spec compliant

> When a promise reject function F is called with argument reason, the following steps are taken:
> 
> 1. Assert: F has a [[Promise]] internal slot whose value is an Object.
> 1. Let promise be the value of F's [[Promise]] internal slot.
> 1. Let alreadyResolved be the value of F's [[AlreadyResolved]] internal slot.
> 1. If alreadyResolved.[[value]] is true, return undefined.
> 1. Set alreadyResolved.[[value]] to true.
>
> https://262.ecma-international.org/6.0/#sec-promise-reject-functions

> When a promise resolve function F is called with argument resolution, the following steps are taken:
> 
> 1. Assert: F has a [[Promise]] internal slot whose value is an Object.
> 1. Let promise be the value of F's [[Promise]] internal slot.
> 1. Let alreadyResolved be the value of F's [[AlreadyResolved]] internal slot.
> 1. If alreadyResolved.[[value]] is true, return undefined.
>
> https://262.ecma-international.org/6.0/#sec-promise-resolve-functions